packages/contract/src/DaimoFastCCTP.sol

```diff
@@ -0,0 +1,241 @@
+  mapping(address => bool) public handoffSent;
+  // On receiving chain, map each transfer to recipient (LP or Bob).
+  mapping(address => address) public handoffToRecipient;
+
+  constructor() {}
+
+  // Called by Alice on Chain A. Requires approval for (fromToken, fromAmount).
+  // Initiates a CCTP transfer.
+  function startTransfer(
+    ICCTPTokenMessenger cctpMessenger,
+    IERC20 fromToken,
+    uint256 fromAmount,
+    uint256 toChainID,
+    address toAddr,
+    uint256 toAmount,
+    IERC20 toToken,
+    uint256 nonce
+  ) public {
+    require(fromAmount >= toAmount, "FCCTP: fromAmount < toAmount");
+
+    // Deploy CCTP sender
+    uint256 fromChainID = block.chainid;
+    address fromAddr = msg.sender;
+    EphemeralHandoff handoff = new EphemeralHandoff{salt: 0}(
+      fromChainID,
+      fromAddr,
+      fromAmount,
+      toChainID,
+      toAddr,
+      toAmount,
+      toToken,
+      nonce
+    );
+
+    // Ensure we don't reuse nonces. This covers the case where Alice sends Bob
+    // the same amount twice in a row = otherwise identical transfers = no way
+    // for an LP to fastFinish() the second one without a distinct handoff addr.
+    require(
+      !handoffSent[address(handoff)],
+      "FCCTP: identical transfer already sent"
+    );
+    handoffSent[address(handoff)] = true;
+
+    // Receive funds from Alice
+    fromToken.safeTransferFrom(fromAddr, address(handoff), fromAmount);
+
+    // Initiate transfer + destroy handoff, minimizing gas consumption.
+    handoff.startTransferAndSelfDestruct(cctpMessenger, fromToken);
+  }
+
+  // Pays Bob immediately on chain B. The caller LP sends (toToken, toAmount).
+  // Later, when the slower CCTP transfer arrives, the LP will be able to claim
+  // (toToken, fromAmount), keeping the spread (if any) between the amounts.
+  function fastFinishTransfer(
+    uint256 fromChainID,
+    address fromAddr,
+    uint256 fromAmount,
+    uint256 toChainID,
+    address toAddr,
+    uint256 toAmount,
+    IERC20 toToken,
+    uint256 nonce
+  ) public {
+    // Calculate handoff address
+    EphemeralHandoff handoff = new EphemeralHandoff{salt: 0}(
+      fromChainID,
+      fromAddr,
+      fromAmount,
+      toChainID,
+      toAddr,
+      toAmount,
+      toToken,
+      nonce
+    );
+    handoff.destroy();
+
+    // Only for transfers which have not already been fastFinished or claimed.
+    require(
+      handoffToRecipient[address(handoff)] == address(0),
+      "FCCTP: already finished"
+    );
+
+    // Record LP as new recipient
+    handoffToRecipient[address(handoff)] = msg.sender;
+
+    // LP pays original recipient
+    toToken.safeTransferFrom(msg.sender, toAddr, toAmount);
+  }
+
+  // Claims a CCTP transfer to its current recipient. If fastFinish... was
+  // called for this transfer, then recipient = the LP who fronted the amount.
+  // Otherwise, the recipient remains the original toAddr, eg Bob.
+  function claimTransfer(
+    ICCTPReceiver cctpReceiver,
```

Comment:
hmm I think you might want to hardcode the cctpReceiver address as a constructor arg (and use create3 to deploy it?)

otherwise I can see some vulnerable edge case around being able to create a fake CCTP receiver that accepts any message/signature and calling this function with those (but not sure)